src/NeoSharp.BinarySerialization/BinaryDeserializer.cs

```diff
@@ -0,0 +1,190 @@
+        /// <param name="type">Type</param>
+        /// <returns>Return object</returns>
+        public object Deserialize(byte[] data, Type type)
+        {
+            using (var ms = new MemoryStream(data))
+            {
+                return Deserialize(ms, type);
+            }
+        }
+        /// <summary>
+        /// Deserialize
+        /// </summary>
+        /// <typeparam name="T">Type</typeparam>
+        /// <param name="stream">Stream</param>
+        /// <returns>Return object</returns>
+        public T Deserialize<T>(Stream stream) where T : new()
+        {
+            // Search in cache
+
+            if (!BinarySerializerCache.Cache.TryGetValue(typeof(T), out var cache))
+                throw new KeyNotFoundException("The type is not registered");
+
+            // Deserialize
+
+            using (var br = new BinaryReader(stream, Encoding.UTF8))
+            {
+                var obj = cache.Deserialize<T>(this, br);
+
+                if (cache.IsOnPostDeserializable)
+                    ((IBinaryOnPostDeserializable)obj).OnPostDeserialize();
+
+                return obj;
+            }
+        }
+        /// <summary>
+        /// Deserialize
+        /// </summary>
+        /// <typeparam name="T">Type</typeparam>
+        /// <param name="stream">Stream</param>
+        /// <returns>Return object</returns>
+        public T Deserialize<T>(BinaryReader stream) where T : new()
+        {
+            // Search in cache
+
+            if (!BinarySerializerCache.Cache.TryGetValue(typeof(T), out var cache))
+                throw new KeyNotFoundException("The type is not registered");
+
+            // Deserialize
+
+            var obj = cache.Deserialize<T>(this, stream);
+
+            if (cache.IsOnPostDeserializable)
+                ((IBinaryOnPostDeserializable)obj).OnPostDeserialize();
+
+            return obj;
+        }
+        /// <summary>
+        /// Deserialize
+        /// </summary>
+        /// <param name="stream">Stream</param>
+        /// <param name="t">Type</param>
+        /// <returns>Return object</returns>
+        public object Deserialize(Stream stream, Type t)
+        {
+            // Search in cache
+
+            if (!BinarySerializerCache.Cache.TryGetValue(t, out var cache))
+                throw new KeyNotFoundException("The type is not registered");
+
+            // Deserialize
+
+            using (var br = new BinaryReader(stream, Encoding.UTF8, true))
+            {
+                var obj = cache.Deserialize(this, br);
+
+                if (cache.IsOnPostDeserializable)
+                    ((IBinaryOnPostDeserializable)obj).OnPostDeserialize();
+
+                return obj;
+            }
+        }
+        /// <summary>
+        /// Deserialize without create a new object
+        /// </summary>
+        /// <param name="buffer">Buffer</param>
+        /// <param name="obj">Object</param>
+        public void DeserializeInside(byte[] buffer, object obj)
```

Comment:
I saw we use heavily *Inside methods. Should we drop Inside suffix for them?

Reply:
What name you propose for Deserialize without create a new object?

Reply:
Could it be another Deserialize method overloading?